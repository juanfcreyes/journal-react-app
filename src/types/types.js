export const types = {
    login: '[Auth] login',
    logout: '[Auth] logout',
    uiSetError: '[UI Set Error]',
    uiRemoveError: '[UI Remove Error]',
    uiStartLoading: '[UI Start Loading]',
    uiFinishLoading: '[UI Finish Loading ]',
    notesAddnew: '[Notes Add New]',
    notesSetActive: '[Notes Set Active]',
    notesLoad: '[Notes Load]',
    notesUploaded: '[Notes Uploaded]',
    notesFileUrl: '[Notes Upload Image Url]',
    notesDelete: '[Notes deleted]',
    notesCleaning: '[Notes cleaning]'
}