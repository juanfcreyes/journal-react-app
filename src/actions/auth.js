import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { finishLoading, startLoading } from "./ui";
import Swal from "sweetalert2";
import { notesLogout } from "./notes";

export const startLogin = (email, password) => {
    return async (dispatch) => {
        dispatch(startLoading());
        return firebase.auth().signInWithEmailAndPassword(email, password).then(({user}) => {
            dispatch(login(user.uid, user.displayName));
            dispatch(finishLoading());
        }).catch((error) => {
            Swal.fire('Error', error.message, 'error');
            dispatch(finishLoading());
            console.log('error', error);
        })
        
    }
}

export const startRegister = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password).then(async ({user}) => {
            await user.updateProfile({displayName: name})
            dispatch(login(user.uid, user.displayName));
        }).catch((error) => {
            Swal.fire('Error', error.message, 'error');
        });
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(({ user }) => {
            dispatch(login(user.uid, user.displayName));
        });
    }
}

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(notesLogout()); 
    }
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
});