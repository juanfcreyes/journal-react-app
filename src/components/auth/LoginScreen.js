import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startLogin, startGoogleLogin } from '../../actions/auth'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {
    
    const ui = useSelector(state => state.ui);
    const dispatch = useDispatch();
        
    const [formValues, handleInputChange] = useForm(
        {
            email:'',
            password: ''
        });

    const {email, password} = formValues;

    const handleLogin = (e) => {
        e.preventDefault(); 
        dispatch(startLogin(email, password));
    }

    const handleGoogleLogin = (e) => {
        e.preventDefault(); 
        dispatch(startGoogleLogin());
    }
    
    return (
        <>
            <h1 className='auth__title'> Login Screen</h1>
            <form onSubmit={handleLogin} className='animate__animated animate__fadeIn'>
                <input className='auth__input' autoComplete='off' type='text' placeholder='email' name='email' value={email} onChange={handleInputChange}/>
                <input className='auth__input' type='password' placeholder='password' name='password' value={password} onChange={handleInputChange}/>
                <button className='btn btn-primary btn-block mb-1 mt-1' type='submit' disabled={ui.loading}>Login</button>
                 <div className='auth__social-networks mb-1'>  
                    <p className='mb-1'>Login with social networks</p>
                    <div className="google-btn" onClick={handleGoogleLogin}>
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link to='/auth/register' className='link'>
                    Create new account
                </Link>
            </form>
        </>
    )
}
