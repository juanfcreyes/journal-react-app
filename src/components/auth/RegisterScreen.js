import React from 'react'
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../../hooks/useForm'
import { Link } from 'react-router-dom'
import { removeError, setError } from '../../actions/ui';
import { startRegister } from '../../actions/auth';

export const RegisterScreen = () => {
    
    const ui = useSelector(state => state.ui); 
    const dispatch = useDispatch();

    const [formValues, handleInputChange] = useForm(
        {
            name: '',
            email:'',
            password1: '',
            password2: ''
        });

    const {name, email, password1, password2} = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startRegister(email, password1, name));
        }
    }

    const isFormValid = () => {

        if (name.trim().length === 0) {
            dispatch(setError('Name is required'));
            return false;
        }

        if (!validator.isEmail(email)) {
            dispatch(setError('Email not valid'));
            return false;
        }

        if (password1 !== password2 || password1.length < 5) {
            dispatch(setError('The password must have more than 5 characters and its confirmation must be the same'));
            return false;
        } 

        dispatch(removeError());

        return true;
    }

    return (
        <>
            <h1 className='auth__title'> Register</h1>
            <form onSubmit={handleRegister}  className='animate__animated animate__fadeIn'>
                {
                    ui.msgError &&
                    <div className='auth__alert-error'>        
                        { ui.msgError }
                    </div>
                }
                <input className='auth__input' autoComplete='off' type='text' placeholder='name' name='name' onChange={handleInputChange}  value={name}/>
                <input className='auth__input' autoComplete='off' type='text' placeholder='email' name='email' onChange={handleInputChange} value={email}/>
                <input className='auth__input' type='password' placeholder='password' name='password1' onChange={handleInputChange} value={password1}/>
                <input className='auth__input' type='password' placeholder='confirm password' name='password2' onChange={handleInputChange} value={password2}/>
                <button className='btn btn-primary btn-block mb-5 mt-1' type='submit'>Register</button>
                <Link to='/auth/login' className='link'>
                   have you already registered?
                </Link>
            </form>
        </>
    )
}
