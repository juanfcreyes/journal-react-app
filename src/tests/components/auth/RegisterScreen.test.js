import { RegisterScreen } from "../../../components/auth/RegisterScreen";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { mount } from "enzyme";
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk';
import { types } from "../../../types/types";

jest.mock('../../../actions/auth', () => ({
    startRegister: jest.fn(),
}));


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('Pruebas sobre RegisterScrenn', () => {

    const initState = {
        ui: {
            loading: false,
            msgError: null
        }
    }

    let store = mockStore(initState);

    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter>
                <RegisterScreen></RegisterScreen>
            </MemoryRouter>
        </Provider>
       );
   
    test('Debe presentarse correctamente', () => {

        expect(wrapper).toMatchSnapshot()        
    });

    test('Debe de hacer el dispatch de la accion', () => {
        const emailField = wrapper.find('input[name="email"]');
        const nameField = wrapper.find('input[name="name"]');
        
        nameField.simulate('change', {
            target: {
                value: 'Juan',
                name: 'name'
            }
        });

        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault: () => {}
        });

        const actions = store.getActions();
        expect(actions[0]).toEqual({type: types.uiSetError, payload: 'Email not valid'})

    });

    test('sDebe de mostrar la alerta con el mensaje', () => {
       
        const state = {
            ui: {
                loading: false,
                msgError: 'El email no es correcto'
            }
        }
    
        const store = mockStore(state);
    
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen></RegisterScreen>
                </MemoryRouter>
            </Provider>
           );
       
        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(state.ui.msgError);
        
    });
    

    
    
});
