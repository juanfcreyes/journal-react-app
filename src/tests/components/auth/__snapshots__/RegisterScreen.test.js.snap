// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre RegisterScrenn Debe presentarse correctamente 1`] = `
Array [
  <h1
    className="auth__title"
  >
     Register
  </h1>,
  <form
    className="animate__animated animate__fadeIn"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="email"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="password1"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5 mt-1"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      have you already registered?
    </a>
  </form>,
]
`;
